build-system/compile/compile.js

```diff
@@ -167,249 +225,232 @@ function compile(
-          compilerOptionsArray.push('--' + option);
-        }
+        compilerFlags.push('--' + option);
       }
-    });
-
-    if (options.typeCheckOnly) {
-      return gulp
-        .src(srcs, {base: '.'})
-        .pipe(sourcemaps.init())
-        .pipe(preClosureBabel())
-        .on('error', (err) => handlePreClosureError(err, outputFilename))
-        .pipe(gulpClosureCompile(compilerOptionsArray))
-        .on('error', (err) => handleTypeCheckError(err))
-        .pipe(nop())
-        .on('end', resolve);
-    } else {
-      timeInfo.startTime = Date.now();
-      return gulp
-        .src(srcs, {base: '.'})
-        .pipe(sourcemaps.init())
-        .pipe(preClosureBabel())
-        .on('error', (err) =>
-          handlePreClosureError(err, outputFilename, options, resolve)
-        )
-        .pipe(gulpClosureCompile(compilerOptionsArray))
-        .on('error', (err) =>
-          handleCompilerError(err, outputFilename, options, resolve)
-        )
-        .pipe(rename(`${outputDir}/${outputFilename}`))
-        .pipe(
-          gulpIf(
-            !argv.pseudo_names && !options.skipUnknownDepsCheck,
-            checkForUnknownDeps()
-          )
-        )
-        .on('error', reject)
-        .pipe(postClosureBabel())
-        .pipe(sanitize())
-        .pipe(writeSourcemaps(options))
-        .pipe(gulp.dest('.'))
-        .on('end', resolve);
     }
   });
+  const jsFlags = globby.sync(transformedSrcFiles).map((src) => `--js=${src}`);
+  const entryPointFlags = entryModuleFilenames
+    .map((entryFile) =>
+      transformedSrcFiles.find((srcFile) => srcFile.endsWith(entryFile))
+    )
+    .map((entryPoint) => `--entry_point=${entryPoint}`);
+  const flags = compilerFlags.concat(entryPointFlags, jsFlags);
+  if (!options.typeCheckOnly) {
+    const outputFileFlag = `--js_output_file=${destFile}`;
+    const sourcemapFlag = `--create_source_map=${sourcemapFile}`;
+    flags.push(outputFileFlag, sourcemapFlag);
+  }
+  return flags;
+}
+
+/**
+ * @param {string[]|string} entryModuleFilenames
+ * @param {string} outputDir
+ * @param {string} outputFilename
+ * @param {!OptionsDef} options
+ * @param {{startTime?: number}} timeInfo
+ * @return {Promise<void>}
+ */
+async function compile(
+  entryModuleFilenames,
+  outputDir,
+  outputFilename,
+  options,
+  timeInfo
+) {
+  if (timeInfo) {
+    timeInfo.startTime = Date.now();
+  }
+  if (!(entryModuleFilenames instanceof Array)) {
+    entryModuleFilenames = [entryModuleFilenames];
+  }
+  const destFile = `${outputDir}/${outputFilename}`;
+  const sourcemapFile = `${destFile}.map`;
+  const compilerOptions = generateCompilerOptions(
+    outputDir,
+    outputFilename,
+    options
+  );
+  const srcs = getSrcs(
+    entryModuleFilenames,
+    outputDir,
+    outputFilename,
+    options
+  );
+  const transformedSrcFiles = await Promise.all(
+    globby.sync(srcs).map(async (src) => {
+      return await preClosureBabel(src, outputFilename, options);
+    })
```

Comment:
Should be functionally the same:
```suggestion
    globby.sync(srcs).map((src) => {
      return preClosureBabel(src, outputFilename, options);
    })
```

Reply:
If this is updated as suggested, might I also nit recommend.

`globby.sync(srcs).map((src) => preClosureBabel(src, outputFilename, options))`

Reply:
Also, in the future it might make sense to see if these babel invocations can be parallelized using something like https://github.com/kristoferbaxter/async with Workers.

Reply:
They should be fully parallelizable!

Reply:
Applied both excellent suggestions.

And yes, the idea is that once this PR lands, we can go in and deliberately make smaller, more targeted changes to speed things up in different places (parallelize where possible, consolidate steps, trim flags / options, etc.)